utility/doc/PROTOCOL_CLIENT_SIDE_CHALLENGE.md

```diff
@@ -0,0 +1,213 @@
+
+Given that we have access to up to $S$ independent blockchain API clients,
+across those $S$ components, we can only ever expect our session to be, at
+best, ***eventually consistent***. While an individual blockchain API client
+will independently remain ***strongly consistent***, since the state propagates
+through the network via peer-to-peer communications, a collection of these
+nodes forms an ***eventually consistent*** system in aggregate. At worst
+however, depending on the underlying logic of the exposed blockchain API
+method, its likely that the session offers ***no consistency***.
+
+### Client Side Challenge Requirements
+
+At minimum, client side challenges can only occur on methods that are
+***idempotent*** and ***eventually consistent*** across a group of nodes.
+
+The underlying response being ***eventually consistent*** means that nodes who
+respond with an error response that indicates the desired resource could not be
+found must be assumed to be operating as expected without further logic
+defining SLA expectations of individual methods.
+
+The **idempotency** requirement of the challenge is to prevent the following
+situations from happening when a node is operating in good faith:
+
+- DoS through chain bloat from an app submitting thousands of valid challenge
+  transactions.
+- Attacks on other node stakes in the situation that client side challenges
+  result in a net loss for the nodes who fail them.
+- Illegitimate reward farming for apps in the situation that client side
+  challenges result in a net gain for the app that executes the challenge.
+
+Additionally, to remain non-interactive, it needs to be the case that the
+relay request is inferable from the result of the relay response.
+
+### Client Side Challenge Mechanism
+
+At the time of whitelisting a relay chain for rewards, it will be required to
+able provide a pattern string that will pattern match to identify if a request
+is unable to be client-side challenged.
+
+This will require that the `pocketcore/SupportedBlockchains` parameter take
+the following form:
+
+```go
+type SupportedBlockchains struct {
+   Blockchains []SupportedBlockchain
+}
+
+type SupportedBlockchain struct {
+    RelayChainID string
+    NotChallengablePatterns []string
+}
+```
+
+The `NotChallengablePatterns` field is an array of pattern strings that should
+successfully match against the raw request string should if the request is not
+able to be client side challenged. This means having to make a change to what
+is expected of the app when requesting a Relay so that the App will send along
+a signed version of the string encoded HTTP request.
+
+```go
+type QueryRelayRequest stuct {
+    Payload       RelayPayload
+    SignedPayload string
+    Meta          RelayMetadata
+    Proof         RelayProof
+}
+```
+
+Then what is expected of the node when responding to a relay is the following
+`RelayResponse`, where the `Signature` field is now the `Signature` of the
```

Comment:
This is a good opportunity to use an equation / pseudo-code instead of text:

```go
RelayResponse.Signature = `Signature(concat(RelayRquest.SignedPayload, RelayResponse.ResponsePayload))
```